src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1537,4 +1536,105 @@ impl Cage {
             }
         }
     }
+
+    //------------------FTRUNCATE SYSCALL------------------
+    
+    pub fn ftruncate_syscall(&self, fd: i32, length: usize) -> i32 {
+        let fdtable = self.filedescriptortable.read().unwrap();
+ 
+        if let Some(wrappedfd) = fdtable.get(&fd) {
+            // check if arg length is valid
+            if length < 0 { 
+                return syscall_error(Errno::EINVAL, "ftruncate", "The argument length is negative");
+            }
+            
+            let filedesc_enum = wrappedfd.read().unwrap();
+            let mut mutmetadata = FS_METADATA.write().unwrap();
+
+            match &*filedesc_enum {
+                // only proceed when fd references a regular file
+                File(normalfile_filedesc_obj) => {
+                    let inodenum = normalfile_filedesc_obj.inode;
+                    let inodeobj = mutmetadata.inodetable.get_mut(&inodenum).unwrap();
+
+                    match inodeobj {
+                        // only proceed when inode matches with a file
+                        Inode::File(ref mut normalfile_inode_obj) => {
+                            // get file object table with write lock
+                            let mut fobjtable = FILEOBJECTTABLE.write().unwrap();
+                            
+                            let mut fileobject = fobjtable.get_mut(&inodenum).unwrap();
+                            let filesize = normalfile_inode_obj.size;
+                            
+                            // if length is greater than original filesize,
+                            // file is extented with null bytes
+                            if filesize < length {
+                                let blankbytecount: usize = length - filesize;
+                                
+                                // create a null-byte buffer of proper size
+                                let mut vec = vec![0u8; blankbytecount];
+                                let buf: *mut u8 = &mut vec[0];
+
+                                // extend file with null-byte buffer
+                                if let Ok(byteswritten) = fileobject.writeat(buf, blankbytecount, filesize) {
+                                    // update filesize if writeat() succeeded
+                                    normalfile_inode_obj.size = length;
+                                } else {
+                                    panic!("writeat() failed");
+                                }
```

Comment:
We made an easier function to do this, called zerofillat, you can look at the pwrite syscall to see its usage